### ISSUE TITLE
[Research] 视频后训练加速 × MiniMax-H3：论文技术映射、ROI 排序与实验计划

### ISSUE THREAD

**Issue by leviking98z-rgb** [leviking98z-rgb/UniRL#42]
## 背景

本 issue 跟踪近期视频生成 / diffusion RL 后训练加速工作在 **UniRL MiniMax-H3** 上的可迁移性、预期上限与验证计划。

它接续 #35 的系统侧 profiling 与 #39 的 SP-group video VAE decode 分片，但不重复已有的 prompt cache、native SP、conditioner sharing 和 decode 优化。重点转向三类仍可能有较高 ROI 的方向：

1. 减少需要 replay/backward 的 timestep；
2. 复用多条 rollout 的共同 denoise prefix；
3. 用低精度探索或 latent reward 降低候选生成 / reward 成本。

相关已有工作：

- 本 fork：#35、#36、#37、#38、#39
- upstream trainside systems roadmap：[Tencent-Hunyuan/UniRL#483](https://github.com/Tencent-Hunyuan/UniRL/issues/483)
- upstream timestep scheduler RFC：[Tencent-Hunyuan/UniRL#488](https://github.com/Tencent-Hunyuan/UniRL/issues/488)
- upstream MiniMax-H3 critical path：[Tencent-Hunyuan/UniRL#391](https://github.com/Tencent-Hunyuan/UniRL/issues/391)
- upstream on-policy distillation roadmap：[Tencent-Hunyuan/UniRL#368](https://github.com/Tencent-Hunyuan/UniRL/issues/368)
- upstream async rollout tracking：[Tencent-Hunyuan/UniRL#94](https://github.com/Tencent-Hunyuan/UniRL/issues/94)

## H3 control 与已知瓶颈

后续实验统一以当前 production-reward workload 为 control：

- MiniMax-H3 33B joint video/audio DiT
- 1×8 H20，full FSDP + SP8
- `batch_size=8`、`samples_per_prompt=4`，每个 rollout 共 32 个完整视频+音频样本
- `768×768×124`，约 22k packed rows
- 10 denoise steps，CPS，`eta=0.7`
- `sde_indices=[0,3,6]`
- FlowGRPO，`old_logp_source=replay`，每 rollout 2 次 optimizer update
- 真实 reward：`0.5 × VideoPickScore(middle frame) + 0.5 × CLAP`

#39 的一次真实 reward / backward A/B 已把最明显的系统瓶颈消掉：

| metric | decode shard off | decode shard on | result |
|---|---:|---:|---:|
| step time | 3536.2 s | **2160.7 s** | **−38.9%** |
| generate time | 2513.3 s | **1128.1 s** | −55.1% |
| video decode / sample | 51.47 s | **8.22 s** | **6.26×** |
| denoise / sample | 19.40 s | 19.39 s | unchanged |

两臂的 32 个完整视频 SHA-256 全部一致。该结果仍是单 rollout A/B，不当作置信区间；它的用途是给下一阶段做 Amdahl 上限判断。

此前在同一正式 workload 上，conditioner sharing 只有约 `0.77%`，next-batch prompt prefetch 约 `1.80%`。因此下一轮不再优先优化 conditioner，而先量清 replay/backward、reward 和 optimizer 各自占比。

## 论文技术映射

| 工作 | 论文中的核心机制 / 报告结果 | 对 H3 的判断 | 优先级 |
|---|---|---|---|
| [Flash-GRPO](https://arxiv.org/abs/2605.15980) ([code](https://github.com/Shredded-Pork/Flash-GRPO)) | 每个 prompt group 只在一个 timestep 做 SDE 与 policy optimization；iso-temporal grouping + temporal gradient rectification。论文在 Wan2.1 1.3B/14B 上报告训练成本约 6× 加速 | **最值得先做。** H3 已从完整轨迹缩到 3 个 replay step，因此收益是 `3→1`，不是把论文 6× 原样搬过来；还必须重推 CPS 与 joint A/V 下的 rectification | **P1** |
| [TreeGRPO](https://arxiv.org/abs/2512.08153) ([code](https://github.com/zh-ding/TreeGRPO)) | 共享 deterministic prefix，在 SDE window 分支；leaf reward 向树根回传形成 per-edge advantage。SD3.5-M 上 iteration `173.5→72.0 s`，约 2.4× | H3 的 4 个 sibling 已共享初始噪声，存在 prefix reuse 的结构机会；但当前 step 0 就是 SDE，必须改变探索窗口，并新增 tree/DAG segment 与 per-edge advantage。收益潜力高、工程风险也高 | **P1/P2** |
| [TempFlow-GRPO](https://arxiv.org/abs/2508.04324) ([code](https://github.com/Shredded-Pork/TempFlow-GRPO)) | trajectory branching、seed group、noise-aware timestep weighting；其任务中早期 timestep reward variance 更大。K=10 时 sampling 平均约为 FlowGRPO 的 4.5×，但达到目标性能只用约 33%–50% 总训练时间 | 不应直接假设 H3 也“越早越重要”。先把它作为逐 timestep probe 与 scheduler 设计依据；实现入口与 upstream #488 重叠 | **P1 诊断 / P2 算法** |
| [MixGRPO](https://arxiv.org/abs/2507.21802) | SDE sliding window、窗口外 ODE；论文报告相对 DanceGRPO 训练时间近 50% 降低，Flash variant 降低 71% | H3 的 `[0,3,6]` + 其余 ODE 已采用这一类 timestep-subset 思路。新增价值主要来自更好的 scheduler，而不是再实现一遍 mixed ODE/SDE | **已有基础能力** |
| [Sol-RL](https://arxiv.org/abs/2604.06916) | 96 个 NVFP4 / 6-step proxy candidates，选 top/bottom 24，再用 BF16 / 10-step 精确重生成并训练；B200 上 rollout `2.33–2.41×`、E2E `1.61–1.62×`，收敛最多 4.64× | 当前 H20 没有论文依赖的原生 NVFP4 路径；可先研究 FP8 proxy。H3 当前 group size 只有 4，且视频 decode + composite reward 很重，只有扩大 candidate pool 后才有意义 | **P2，硬件/系统受限** |
| [DigenRL](https://arxiv.org/abs/2606.24369) | disaggregated train/rollout、generation-axis pipeline、timestep parallelism、trainer-assisted generation；论文报告 1.56–2.10× throughput | fork 的 [`feat/digenrl-tsp`](https://github.com/leviking98z-rgb/UniRL/tree/feat/digenrl-tsp) 已有 AsyncDiffusionTrainer 原型，但 H3 trainside 不能直接跨 slab；要先完成 H3 dedicated rollout engine、weight sync 与 #391 critical path。论文当前标记为 withdrawn，数字只作线索 | **条件性 P2** |
| [PRFL / Video Generation Models Are Good Latent Reward Models](https://arxiv.org/abs/2511.21541) | 用 video DiT 前 8 个 blocks + query attention / MLP 在 noisy latent 上评分，不做 VAE decode；相对 first-frame RGB ReFL 报告 `1.42–1.49×` | 这是新的 differentiable ReFL/latent-reward 路线，不是当前 terminal-reward FlowGRPO 的等价插件；还需要专门偏好标注。它只能先替代 video reward，CLAP 仍需要 audio decode | **P2 研究项** |

### 为什么优先

它最贴合 H3 当前接口：完整 10-step rollout 仍用于得到真实终局 reward，只把需要存轨迹、replay 和 backward 的 transition 从 3 个缩到 1 个。

但它不会减少 10 次 rollout DiT forward，也不会减少 decode/reward。用 #39 的结果做一个极端乐观上限：假设 `step−generate = 2160.7−1128.1 = 1032.6 s` **全部**都能随 replay step 数线性缩放，则：

```text
new_step >= 1128.1 + 1032.6 / 3 = 1472.3 s
speedup <= 2160.7 / 1472.3 = 1.47×
```

实际值必然低于 `1.47×`，因为 reward、optimizer、通信和固定开销不会随 `3→1` 一起缩小。论文的约 6× 不能表述为 H3 预期结果。

### H3 适配点

1. 当前 `DiffusionSamplingParams.resolve_sde_indices(rollout_id)` 每个 rollout 只解析一次，`LatentSegment.sde_indices` 也是 shared field；这只能让整个 rollout 共用一个 timestep。完整 iso-temporal grouping 要做到“同 prompt 的 4 个 sibling 同 timestep、不同 prompt 可选不同 timestep”，需要 group-local timestep metadata。
2. H3 当前 `init_same_noise=true`，siblings 共享 `x_T`，主要由所选 SDE transition 的 per-sample noise 产生差异；这与论文实验的初始化约定不同，需单独验证 advantage spread。
3. Flash-GRPO 的 `λ(t)` 从其 FlowSDE transition 推导。H3 使用 CPS：`std = sigma_next × sin(etaπ/2)`，并有独立 video/audio sigma grid，不能复制论文公式。
4. `audio_joint_sde=true` 时，H3 把 video/audio log-prob 按 latent element 数加权。rectification 必须保持这个联合 policy 语义，不能无意中改变 A/V 权重。

### 实验矩阵

- baseline：`sde_indices=[0,3,6]`
- diagnostic：静态 `[0]`、`[3]`、`[6]`
- rollout-level random one-step：先复用当前 shared metadata，验证吞吐、数值和恢复路径
- exact Flash：prompt-group iso-temporal one-step + CPS temporal rectification
- 对每个 timestep 记录 raw/corrected gradient norm、reward variance、advantage magnitude、clip fraction 和样本质量
- 最终按相同 GPU-hours 比较，不只比较单 step latency 或 training reward

静态 one-step / naive random one-step 只作诊断；论文已经显示没有 rectification 的单步训练可能失稳，不应直接作为长跑 recipe。

## P1/P2：Tree prefix reuse 与 tree advantage

先看一个只计算 denoise forward 的简化上限。4 个 sibling、10 steps 独立生成需要 `40` 次 DiT forward；若前 `p` 步确定性共享、之后才分成 4 个 leaf，则：

```text
40 → p + 4 × (10 − p)

p=3: 40 → 31，denoise-only 1.29×
p=6: 40 → 22，denoise-only 1.82×
```

这不是 E2E 加速：4 个 leaf 仍要各自 decode 和打 reward，而且 H3 当前 `[0,3,6]` 在 step 0 已分支，没有可直接复用的 deterministic prefix。

建议拆成两层：

1. **prefix-reuse smoke**：branch factor 2、两层小树，固定 4 个 terminal leaves；先验证 forward count、leaf seed、terminal output、峰值显存和 SP8 collective 顺序。
2. **完整 TreeGRPO**：增加 parent/edge id、branch timestep、old log-prob、leaf-to-root backup 和 per-edge advantage；对照 vanilla terminal advantage。

主要风险：H3 目前 `forward_batch_size=1`，约 22k rows，并要求 SP8 lockstep。不能直接假设论文在 SD3.5-M 上的 frontier batching 和 2.4× 能复现；树的 frontier 若并行展开，还可能把当前不足 1 GiB 的显存余量耗尽。

## P2：低精度大规模探索

Sol-RL 的思想值得保留，但 H3 上应先做 **ranking audit**，不是直接实现大规模训练：

1. 用完全相同的 initial-noise / SDE-noise seed，生成 FP8 proxy 与 BF16 control；
2. 分别检查 VideoPickScore、CLAP 和 composite reward 的 Spearman/Kendall correlation；
3. 检查 Top/Bottom-K 命中率，而不只看平均 reward；
4. 验证选中 seed 能被 BF16 精确重生成，并且 replay anchor 来自 BF16 policy；
5. 从 `N=16,K=4` 小规模开始，确认 decode/reward 没有吞掉 DiT 量化收益，再考虑 `N=32/64/96`。

若没有 dedicated rollout engine、稳定 weight sync 和量化权重刷新，H3 trainside 上不应优先投入。fork 的 `feat/fp8-rollout` 是 AR/SGLang 路径，不等于 H3 diffusion rollout 已支持 FP8。

## P2：latent video reward / PRFL

这条线的潜在价值不仅是省 decode，也可能补上当前 VideoPickScore 只看中间帧、无法评价运动的缺陷。但它改变了 reward 和优化目标，应作为独立研究项目：

- 收集 H3 视频偏好 / motion-quality 标签；
- 训练 H3-compatible latent video reward，并做 noisy-timestep calibration；
- 对当前 VideoPickScore、VBench、人工偏好和运动指标做相关性验证；
- 先在 frozen rollout 上离线评估，再考虑 ReFL/PRFL 训练；
- 保留 CLAP 或另训 latent audio reward，否则仍需 audio decode。

#39 已把 video decode 从 `51.47 s/sample` 降到 `8.22 s/sample`，所以 PRFL 在 H3 上的增量加速上限会明显低于论文相对 RGB ReFL 的数字；是否采用应更多由 reward 质量而非纯吞吐决定。

### TeaCache

[TeaCache](https://arxiv.org/abs/2411.19108) 在视频 diffusion inference 上报告最高 4.41×，但主 on-policy trajectory 中跳过 DiT forward 会改变 behavior policy 与 rollout/replay likelihood。H3 只有 10 steps，且每个 rollout 后更新权重，cache 的可复用窗口也短。

结论：可用于 eval，或作为“低精度 proxy → 精确重生成”体系的一部分；不直接用于主 FlowGRPO rollout。

### Diffusion Adversarial Post-Training

[APT](https://arxiv.org/abs/2501.08316) 能把视频生成蒸馏到一步，但需要真实数据、discriminator 和完整 adversarial post-training。它是新 actor/model 训练项目，不是现有 H3 FlowGRPO loop 的加速开关。

## Phase 0：先补可归因 telemetry

在任何算法 A/B 前，先把一个 step 拆成：

- conditioner / denoise / video decode / audio decode
- VideoPickScore / CLAP
- old-logp anchor replay
- update replay + backward（按 update、按 timestep）
- optimizer / offload-onload / checkpoint

同时记录每个 selected timestep 的 video/audio sigma、reward variance、advantage、raw/corrected gradient norm、clip fraction、peak memory 和 DiT forward count。没有这组数据，就无法区分“算法没有收益”和“收益被 decode/reward/optimizer 固定成本遮住”。

## 推荐顺序

1. **P0：phase instrumentation + per-timestep probe**
2. **P1：Flash-GRPO rollout-level one-step smoke**
3. **P1：prompt-group iso-temporal metadata + CPS rectification**
4. **P1/P2：TreeGRPO 小树 / prefix-reuse smoke**
5. **P2：FP8 proxy ranking audit**
6. **P2：latent video reward feasibility**
7. **条件满足后：DigenRL / disaggregated async H3**

## GO / NO-GO 标准

任何方案在标记 GO 前至少满足：

- production VideoPickScore+CLAP，非零、非退化 advantage / gradient；
- checkpoint save + fresh restore 后继续 forward/backward/update；
- 多 seed；性能比较尽量使用 AB/BA crossover；
- 报告完整 GPU-hours、samples-to-quality 和 E2E wall time，不只报告局部 kernel 或单 step；
- held-out reward、VBench/运动指标、人工抽检、KL/多样性无显著退化；
- 在质量持平前提下，E2E 改善至少约 10% 才进入生产化，容量或显著质量收益可单独例外讨论；
- 论文数字、H3 理论上限和 H3 实测结果必须分栏，不把跨模型结果写成已复现结论。

## Checklist

- [ ] 补齐 phase / per-timestep telemetry
- [ ] 完成 `[0]` / `[3]` / `[6]` one-step correctness 与 timing probe
- [ ] 实现 rollout-level random one-step baseline
- [ ] 设计 prompt-group-local `sde_indices` 数据契约
- [ ] 推导并数值校验 CPS joint video/audio temporal rectification
- [ ] 完成 Flash-GRPO 多 seed、相同 GPU-hours 对照
- [ ] 完成 4-leaf TreeGRPO prefix-reuse smoke
- [ ] 验证 per-edge advantage 与 leaf-to-root backup
- [ ] 完成 FP8/BF16 reward ranking audit
- [ ] 评估 H3 latent video reward 数据与标注成本
- [ ] 仅在 H3 dedicated rollout engine 可用后重启 disaggregated async 实验

AI assistance：本调研与 issue 草稿使用 Codex 辅助；论文报告数字、H3 已测数字和推导上限已明确分开，后续结果继续更新到本 issue。
